util/rewrite_all_hw.py

```diff
@@ -0,0 +1,310 @@
+    # Use ripgrep to find all matching files
+    res = subprocess.run(
+        ["rg", "-l", search],
+        capture_output = True,
+        cwd = project_root,
+    )
+    # ripgrep returns 1 if there are no matches, 2 on error
+    if res.returncode == 1:
+        return []
+    assert res.returncode == 0, "ripgrep command failed"
+
+    return [Path(os.fsdecode(path)) for path in res.stdout.splitlines()]
+
+
+def global_replace(project_root, search, replace, verbose):
+    print(f'global replace "{search}" by "{replace}"')
+    for path in find_all_files(project_root, search):
+        path = project_root / path
+        if verbose:
+            print(f"Patching {path}")
+        # Read, patch, write
+        f = path.read_text()
+        f = f.replace(search, replace)
+        path.write_text(f)
+
+
+def run_buildifier(project_root):
+    subprocess.run(
+        ["./bazelisk.sh", "run", "//quality:buildifier_fix"],
+        check=True,
+        cwd = project_root
+    )
+
+
+def run_topgen(project_root):
+    for top in all_tops:
+        subprocess.run(
+            ["./util/topgen.py", "-t", f"hw/top_{top}/data/top_{top}.hjson"],
+            check=True,
+            cwd = project_root,
+        )
+
+    subprocess.run(
+        ["make", "-C", "hw", "cmdgen"],
+        check=False,
+        cwd = project_root,
+    )
+
+
+def delete_rule(lines, rule_name, target_name, file_name):
+    try:
+        start_idx = 0
+        while start_idx < len(lines):
+            start_idx = lines.index(f'{rule_name}(\n', start_idx)
+            if target_name is None or lines[start_idx + 1] == f'    {target_name},\n':  # noqa: E231
+                break
+            start_idx = start_idx + 1
+    except ValueError:
+        assert False, \
+            f"did not find beginning of rule {rule_name} (target name {target_name}) in {file_name}"
+    try:
+        end_idx = lines.index(')\n', start_idx + 1)
+    except ValueError:
+        assert False, \
+            f"did not find end of rule {rule_name} (target name {target_name}) in {file_name}"
+
+    if start_idx > 0 and lines[start_idx - 1] == '\n' and \
+            end_idx + 1 < len(lines) and lines[end_idx + 1] == '\n':
+        end_idx += 1
+
+    return lines[:start_idx] + lines[end_idx + 1:], start_idx, lines[start_idx:end_idx + 1]
+
+
+def delete_all_rules(lines, rule_name, target_name, file_name):
+    while True:
+        try:
+            lines, _, _ = delete_rule(lines, rule_name, target_name, file_name)
+        except:  # noqa: E722
+            break
+    return lines
+
+
+def step1(project_root, commit):
+    new_files = []
+    for (_ippath, info) in all_ips.items():
+        ippath = project_root / Path(_ippath)
+        ip_name = ippath.name
+        is_template = info.get("is_template", False)
+
+        def_file_path = ippath / ("defs.bzl.tpl" if is_template else "defs.bzl")
+        # build_file_path = ippath / ("BUILD.tpl" if is_template else "BUILD")
```

Comment:
Yes some (all?) ipgen IP template `BUILD` file. Well technically they are not templates because their content does not depend on the template but we can't name them just `BUILD` because then bazel considers them as a package and this can cause all sorts of weird issues potentially. Therefore calling `BUILD.tpl` seems like an easy way to avoid this issue. Unless you have a better idea on how to handle this?

Reply:
Will we be able to dispense with the "all_files" target? That may make these BUILD and BUILD.tpl files unnecessary.

Reply:
Unfortunately, there was quite a lot of pushback against touching the "all_files" targets so I have given up on touching them for the time being.